app-engine/java/src/com/google/cloud/tools/intellij/appengine/java/inspections/AppEngineDeprecatedRuntimeInspection.java

```diff
@@ -0,0 +1,121 @@
+ * values. Deprecated values will be underlined with an error, and a quickfix to update to the
+ * supported Java 8 runtime will be presented
+ */
+public class AppEngineDeprecatedRuntimeInspection extends XmlSuppressableInspectionTool {
+
+  private List<String> deprecatedRuntimes = ImmutableList.of("java", "java7");
+
+  private static final String APP_ENGINE_JAVA8_RUNTIME_VALUE = "java8";
+  private static final String APP_ENGINE_WEB_XML_ROOT_TAG_NAME = "appengine-web-app";
+  private static final String APP_ENGINE_WEB_XML_RUNTIME_TAG_NAME = "runtime";
+
+  @Nls
+  @NotNull
+  @Override
+  public String getDisplayName() {
+    return AppEngineMessageBundle.message(
+        "appengine.deprecated.java.runtime.inspection.display.name");
+  }
+
+  @Nullable
+  @Override
+  public String getStaticDescription() {
+    return AppEngineMessageBundle.message(
+        "appengine.deprecated.java.runtime.inspection.description");
+  }
+
+  @NotNull
+  @Override
+  public PsiElementVisitor buildVisitor(@NotNull ProblemsHolder holder, boolean isOnTheFly) {
+    return new XmlElementVisitor() {
+
+      @Override
+      public void visitXmlTag(XmlTag tag) {
+        if (isAppEngineWebXmlDeprecatedRuntimeTag(tag)) {
+          holder.registerProblem(
+              tag,
+              AppEngineMessageBundle.message(
+                  "appengine.deprecated.java.runtime.inspection.problem.text"),
+              new UpdateDeprecatedAppEngineJavaRuntimeQuickFix());
+        }
+      }
+    };
+  }
+
+  /**
+   * A quickfix that updates deprecated App Engine Java runtime values to the supported Java 8
+   * runtime.
+   */
+  private class UpdateDeprecatedAppEngineJavaRuntimeQuickFix implements LocalQuickFix {
+
+    @Nls(capitalization = Capitalization.Sentence)
+    @NotNull
+    @Override
+    public String getFamilyName() {
+      return AppEngineMessageBundle.message("appengine.deprecated.java.runtime.quickfix.text");
+    }
+
+    @Override
+    public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor descriptor) {
+      XmlTag xmlTag = (XmlTag) descriptor.getPsiElement();
+      xmlTag.getValue().setText(APP_ENGINE_JAVA8_RUNTIME_VALUE);
+    }
+  }
+
+  /**
+   * Returns {@code true} if the inspected tag is part of an appengine-web.xml config file, is a
+   * runtime tag, and the value is a deprecated Java runtime.
+   */
+  private boolean isAppEngineWebXmlDeprecatedRuntimeTag(XmlTag tag) {
+    XmlFile xmlFile = (XmlFile) tag.getContainingFile();
+    boolean isAppEngineWebXml =
+        xmlFile.getRootTag() != null
+            && AppEngineUtil.APP_ENGINE_WEB_XML_NAME.equals(xmlFile.getName())
+            && APP_ENGINE_WEB_XML_ROOT_TAG_NAME.equals(xmlFile.getRootTag().getName());
+
+    if (isAppEngineWebXml) {
+      return APP_ENGINE_WEB_XML_RUNTIME_TAG_NAME.equals(tag.getName())
+          && deprecatedRuntimes.contains(tag.getValue().getText());
```

Comment:
Just in case, probably worth checking for `null` `getValue()` here?

Reply:
getValue() is annotated `@NotNull`

Reply:
oh, cool, thanks.